### PR TITLE
SMW: v2.0 Content Update

### PR DESCRIPTION
**Pull request by PoryGone**
Major credit to @TheLX5 for being the driving force for almost all of this update. We've been collaborating on design and polish of the features for the last few months, but all of the heavy lifting was all @TheLX5.

This update has been tested via a couple of controlled public betas in the AP Discord, and has now been available as an APWorld for over a month, and has seen decent play and positive feedback by those channels.

Changelog:

Features:
- New optional Location Checks
  - 3-Up Moons
  - Hidden 1-Ups
  - Bonus Blocks
  - Blocksanity
    - All blocks that contain coins or items are included, with the exception of:
      - Blocks in Top Secret Area & Front Door/Bowser Castle
      - Blocks that are unreachable without glitches/unreasonable movement
- New Items
  - Special Zone Clear
  - New Filler Items
    - 1 Coin
    - 5 Coins
    - 10 Coins
    - 50 Coins
  - New Trap Items
    - Reverse Trap
    - Thwimp Trap
- SFX Shuffle
- Palette Shuffle Overhaul
  - New Curated Palette can now be used for the Overworld and Level Palette Shuffle options
  - Foreground and Background Shuffle options have been merged into a single setting
- Max possible Yoshi Egg value is 255
  - UI in-game is updated to handle 3-digits
  - New `Display Received Item Popups` option: `progression_minus_yoshi_eggs`

Quality of Life:
- In-Game Indicators are now displayed on the map screen for location checks and received items
- In-level sprites are displayed upon receiving certain items
- The Camera Scroll unlocking is now only enabled on levels where it needs to be
- SMW can now handle receiving more than 255 items
- Significant World Code cleanup
  - New Options API
  - Removal of `world: MultiWorld` across the world
- The PopTracker pack now has tabs for every level/sublevel, and can automatically swap tabs while playing if connected to the server

Bug Fixes:
- Several logic tweaks/fixes

### CODE REVIEW

File: worlds/smw/Client.py
Comment (Zunawe) [inline worlds/smw/Client.py:534]: I find these instances where there are no spaces around operators to be cramped, and I'd bet it's not PEP8, but it's not worth blocking the PR over.

File: worlds/smw/Options.py
Comment (Zunawe) [inline worlds/smw/Options.py:340]: This snake case format for option values is inconsistent with at least `SFXShuffle`, which capitalizes the option as it would be on the webhost.

**Review by Zunawe — COMMENTED** [review]
Mostly just scrolled through the code looking for easy stuff. I don't play/own SMW, so I didn't check logic, play games, or run `generate_output`. I also didn't check any asm or data.

Styling/linting stuff doesn't really matter. Just figured I'd point it out since it's there. The new `Regions.py` code occasionally has some extra/missing spaces that I didn't point out.

It'd be nice to have more type hints on more function parameters, but for the most part this is only a lack of change to existing code, and so arguably out of scope.

Generated a few solo seeds and a 100-player seed with a yaml that had everything set to random. No problems. World load and location benchmarks don't stick out as abnormal.

As far as being consistent with the way the world currently works, it seems pretty much fine. Doesn't really shake things up, just adds some new stuff and fixes up a bunch of individual lines.

File: worlds/smw/__init__.py
Comment (Zunawe) [reply worlds/smw/__init__.py:189]: Rats, it didn't looking at this code:
```py
        trap_pool = []
        for i in range(trap_count):
            trap_item = self.multiworld.random.choice(trap_weights)
            trap_pool.append(self.create_item(trap_item))
```

**Comment by Zunawe** [comment]
Whoops, forgot to add a couple more notes:
- Searched for usages of the old options system, couldn't find any
- Only found one use of `multiworld.random`, which I marked
- It's probably worth going through each file and pruning/organizing the imports. Lots of unused stuff.

File: worlds/smw/Options.py
Comment (Zunawe) [reply worlds/smw/Options.py:340]: > created by some people
This is also vague in a sort of silly way, and could maybe be rephrased.

File: worlds/smw/Regions.py
Comment (alwaysintreble) [reply worlds/smw/Regions.py:2160]: can't comment on it specifically since it's old code, but the entire rest of this method can pretty much be replaced with just `source_region.connect(target_region, target_region.name, rule)`

File: worlds/smw/Rom.py
Comment (alwaysintreble) [inline worlds/smw/Rom.py:771]: swap to orjson when you can. it returns bytes, so you may need to rework the way you're reading it a bit.

File: worlds/smw/__init__.py
Comment (alwaysintreble) [reply worlds/smw/__init__.py:88]: it'd be way nicer if it used `self.options.as_dict` as well

File: worlds/smw/__init__.py
Comment (alwaysintreble) [inline worlds/smw/__init__.py:202]:
```suggestion
        junk_pool = [self.create_item(self.random.choice(junk_weights)) for _ in range(junk_count)]
```
maybe?

File: worlds/smw/__init__.py
Comment (alwaysintreble) [inline worlds/smw/__init__.py:193]: none of these parens should be here

File: worlds/smw/__init__.py
Comment (alwaysintreble) [inline worlds/smw/__init__.py:288]: not quite sure if it makes sense in this context but in case it does
```suggestion
                    if not self.options.moon_checks and "3-Up Moon" in loc_name:
```

File: worlds/smw/__init__.py
Comment (alwaysintreble) [reply worlds/smw/__init__.py:333]: the if else is also unnecessary. `key = lambda item: item.name == "Yoshi Egg"` accomplishes the same thing

Comment (alwaysintreble) [reply worlds/smw/__init__.py:333]: on an unrelated note i really hate this but i understand why you're doing it and i don't really have a better solution.

File: worlds/smw/Locations.py
Comment (alwaysintreble) [inline worlds/smw/Locations.py:988]: don't need the `.value`s here

File: worlds/smw/Options.py
Comment (alwaysintreble) [inline worlds/smw/Options.py:335]: since you have curated palettes it might be worth considering making this a TextChoice so people can specify using specific curated palettes instead of being forced to get one at random?

**Review by alwaysintreble — COMMENTED** [review]
mostly minor nitpicks. code overall looks good

File: worlds/smw/Aesthetics.py
Comment (alwaysintreble) [reply worlds/smw/Aesthetics.py:4]: an alternative way to do it is to add an assert everywhere you use it. so `assert isinstance(world, SMWWorld)` which will cast it to the correct type in your ide, and gets dropped when frozen. the way zunawe wrote it is the technically correct way though.

**Review by nbrochu — COMMENTED** [review]
I'll start by saying I have played this content update to goal on 3 separate occasions with most of the new options turned on (blocksanity, moon checks, hidden 1-up checks, bonus blocks, curated palette shuffle, SFX shuffle). Everything I've seen works as expected. I did not test the new traps and yoshi egg hunt. The writing to data storage works as expected too (can be confirmed by using the PopTracker packs and seeing the tabs switch automatically)

Code-wise, chipping in with a few things I noticed in the client (since other people have already covered the rest really well). Nothing that should be blocking this PR.

**Review by nbrochu — APPROVED** [review]

**Review by Zunawe — APPROVED** [review]
All my comments were either fixed, acknowledged, or inconsequential styling nitpicks in the first place.

I'll defer to a core reviewer's opinion on the file loading stuff, but I think it's unlikely to be of major concern as-is anyway.

File: worlds/smw/__init__.py
Comment (Zunawe) [inline worlds/smw/__init__.py:61]: This is largely a question for my own understanding, but worth pointing out here anyway, I suppose:

The next AP version won't ship as version 0.4.4, and we all know this when developing clients that get shipped with AP. But an actual 0.4.4 install can connect to a server running a game that used this version of the world to generate, and AP won't reject it. So if we merge this and immediately ship the next AP version, incompatible clients will be allowed to connect to new slots. But in trying to predict the next release number during one of these PRs you make it impossible to connect your client on the current commit to a game generated on the current commit (and can just be wrong about the version number).

Do we have a solution for this? Is the expectation just to play catch-up with your world's `required_client_version` between releases by making a PR to bump it between the AP version bump and the first RC?

File: worlds/smw/Regions.py
Comment (alwaysintreble) [inline worlds/smw/Regions.py:2170]: this should really be using `Region.connect` instead of `Region.add_exits`
```suggestion
    target_region: Region = world.multiworld.get_region(target, world.player)
    source_region.connect(target_region, rule)
```
 there are also getter helpers on the World now so you can just do `target_region: Region = world.get_region(target)`

**Review by alwaysintreble — APPROVED** [review]
didn't test anything but all the code changes LGTM. i guess i should ask if there should be any doc changes with this update to include the new options but i'm not really sure how the docs are currently setup and that might not make sense.

File: worlds/smw/Locations.py
Comment (Berserker66) [inline worlds/smw/Locations.py:987]: here you are unpacking, then repacking, then updating. You can skip the first two.
```suggestion
        location_table.update(moon_location_table)
```

**Review by Berserker66 — APPROVED** [review]

**Pull request merged**
